### ISSUE TITLE
`count` isn't -9999 but 0 

### ISSUE THREAD

**Issue by vincentsarago** [PDAL/PDAL#2447]
Hi,
First I apologize if this question has already be raised. 

when doing `pdal translate` of a `.laz` file to a GDAL raster using the command above: 
```
pdal translate in.laz out.tif --writers.gdal.dimension=Z --writers.gdal.resolution=0.25
```

The nodata value is set for the whole raster to be `-9999`

```
gdalinfo out.tif
Driver: GTiff/GeoTIFF
...
Band 1 Block=301x3 Type=Float64, ColorInterp=Gray
  Description = min
  NoData Value=-9999
Band 2 Block=301x3 Type=Float64, ColorInterp=Undefined
  Description = max
  NoData Value=-9999
Band 3 Block=301x3 Type=Float64, ColorInterp=Undefined
  Description = mean
  NoData Value=-9999
Band 4 Block=301x3 Type=Float64, ColorInterp=Undefined
  Description = idw
  NoData Value=-9999
Band 5 Block=301x3 Type=Float64, ColorInterp=Undefined
  Description = count
  NoData Value=-9999
Band 6 Block=301x3 Type=Float64, ColorInterp=Undefined
  Description = stdev
  NoData Value=-9999
```

But then when I look at the raster, the value of `count` where I expect to have `nodata` is set at `0`. I understand why the count is 0 ... but from a raster/gdal POV I think it should be set to whatever the `nodata` value is.

Thanks.

**Comment by abellgithub** [comment]
Would you want the NODATA value to be shown as 0?

We use NODATA where we can't determine a value, but for count, we can determine a value for every raster cell.  Sometimes that value is 0.  My intuition would be that for calculations, having a value of 0 is better than some arbitrary NODATA value, though I have little experience with this.

**Comment by vincentsarago** [comment]
Thanks for the quick answer @abellgithub. 

Yeah the count is 0 because there is no lidar return .... thus technically speaking this is `no data`. I understand that changing the value from 0 to `-9999` could break things tho.

> Would you want the NODATA value to be shown as 0?

The problem is that GeoTIFF doesn't support multiple NoData Values.

**Comment by abellgithub** [comment]
I don't think there's a right answer.  I understand your point, and it might be more consistent doing what you suggest so that all bands will have nodata values in the same cell positions.  That said, I'm hesitant to change behavior for something that's not explicitly wrong.  I could see someone using the 0-value to do math or as a sentinel.

Perhaps others will chime in.

**Comment by hobu** [comment]
I think we want to use a [MaskBand](https://trac.osgeo.org/gdal/wiki/rfc15_nodatabitmask) and set NODATA with that instead of using a single NODATA value. The issue with `count ==0` is that is real information, and NULL is not the same as ZERO. 

Not all GDAL drivers support MaskBands, but users can use GDAL to manipulate them as they need with driver options and such.

**Comment by vincentsarago** [comment]
thanks for jumping on this @hobu. MaskBand is an excellent solution but I'm just worry about some edge case I'm seen in the past where MaskBand didn't work well for non byte dataset https://github.com/cogeotiff/rio-cogeo/issues/54 but this need more testing

**Comment by hobu** [comment]
> didn't work well for non byte dataset

You can control the output type with the `data_type` option, which might help. Depending on the scenario, you might want integer grids anyway.

I wonder if we could add an option to `writers.gdal`, which is off by default, that allows you to have the `zeroIsNoData` behavior you want.

**Comment by abellgithub** [comment]
I would think there would be a raster utility that would let you simply replace values, no?

**Comment by hobu** [comment]
> I would think there would be a raster utility that would let you simply replace values, no?

True. Simply doing MaskBands should provide enough capability here.

**Comment by abellgithub** [comment]
@hobu: Wondering what we should do with this?  Is this a problem for raster-land, or something we should support?

**Comment by hobu** [comment]
Adding MaskBand support should be done during the next significant updates to `writers.gdal`. Happy to take a patch on this, but for now we're pushing it to future.
